### PR TITLE
Upgrade slevomat/coding-standard to v6

### PR DESCRIPTION
**Pull request by simPod**
This aggregates:
- https://github.com/doctrine/coding-standard/pull/142
- https://github.com/doctrine/coding-standard/pull/139
- https://github.com/doctrine/coding-standard/pull/138

Things to note:

- There are few `traversableTypehints` listed for type hints related sniffs. Originally, there was only Doctrine Collection.
-  Spaces are not required before block control structures now, it's status quo. It's possible to enforce it by disabling excludes for `BlockControlStructureSpacing` sniff if desired.

### CODE REVIEW

**Comment by lcobucci** [comment]
As I mentioned to @simPod I'll be taking a look at this to ensure we enforce typed properties on PHP 7.4+

**Comment by alcaeus** [comment]
> As I mentioned to @simPod I'll be taking a look at this to ensure we enforce typed properties on PHP 7.4+

Can you elaborate? Will this enforce typed properties if phpcs is run on PHP 7.4, or if the library supports PHP ^7.4?

File: tests/fixed/type-hints.php [hunk 1]
```diff
@@ -0,0 +1,24 @@
...
+
+class TraversableTypeHints
+{
+    /** @var Traversable */
+    private $parameter;
+
+    /**
+     * @param Iterator $iterator
+     *
+     * @return Traversable
```
Comment (alcaeus) [inline tests/fixed/type-hints.php]: How is this type hint still in the fixed file? Shouldn't this have a specification, like `Traversable<string>`?

Comment (simPod) [reply tests/fixed/type-hints.php]: Yes but it can't be automatically fixed, AFAIK that's what fixed folder is for, hm? So after running `phpcbf` it does not change

Comment (alcaeus) [reply tests/fixed/type-hints.php]: Ok, I understand. At first I thought it should remove it and then I realised that it shouldn't because it needs a generic type hint. Just didn't make the final connection, thanks for that!

File: tests/input/example-class.php [hunk 2]
```diff
@@ -122,6 +122,7 @@ public function throwWhenInvalid() : void
     public function trySwitchSpace() : void
     {
         try {
+            $var = 1;
```
Comment (Ocramius) [inline tests/input/example-class.php]: This kinda snuck in: not a problem for me as long as it doesn't remove newlines before control structures, but seems unrelated

Comment (simPod) [reply tests/input/example-class.php]: I just ensured the settings for space before block control structures is not modified from current behaviour. The test for it was missing.

**Comment by kukulich** [comment]
@lcobucci @alcaeus 

https://github.com/slevomat/coding-standard/blob/master/SlevomatCodingStandard/Sniffs/TypeHints/PropertyTypeHintSniff.php#L50

That's the default value but it can be changed in configuration: https://github.com/doctrine/coding-standard/pull/144/commits/9578c548e8a352d2ca75fef6094bfb9471795293#diff-1d794c8c2cf3bf4d562ddaea8be99bb0R349

**Comment by alcaeus** [comment]
Thanks @kukulich! Just wanted to make sure it remains optional.

**Comment by carusogabriel** [comment]
Let's wait to merge this one, we want to perform some functional tests across our repos to see if everything is okay 👍

**Comment by alcaeus** [comment]
@carusogabriel here's ODM: https://gist.github.com/alcaeus/29f2c69ca1444260ea2b976221942c5c

**Comment by simPod** [comment]
@alcaeus you should probably upgrade also these excludes https://github.com/doctrine/mongodb-odm/blob/3ac6f4596efce4fb4ff2feb85267d34ddb403c35/phpcs.xml.dist#L28-L29

**Comment by alcaeus** [comment]
@lcobucci can we please split out the patch for PHP 7.4? I have my doubts about the methodology and would rather not have it block this PR. Thanks!

**Comment by alcaeus** [comment]
@simPod thanks for the hint, totally missed those. Gist is updated - most are BlockControlSpacing with a few missing parameter type hints where we either need to add documentation or update the exclude list. At first glance, it looks good 👍

**Comment by lcobucci** [comment]
> I have my doubts about the methodology and would rather not have it block this PR. Thanks!

What are your doubts? Could we look at them and discuss before that?

**Comment by alcaeus** [comment]
Happy to do so in a separate PR. Apart from PHP 7.4, this was ready to go 2 hours ago apart from checking our repos. We split work into multiple PRs so we can vote and discuss individually.

**Comment by lcobucci** [comment]
My main point here is that I'd much rather rely on default sniff behaviour than having it done via configuration (which was the previous solution and it's a no-go from me).

If you have doubts about the solution, share them.

**Comment by alcaeus** [comment]
Mainly, tooling and documentation around using patch files to test behaviour on PHP 7.4. Also, implicitly requiring type hints when running on PHP 7.4 - I don't consider that a sane default. Until we require PHP 7.4 for our projects, `false` should be the default for that property.

Apart from that, this PR was about upgrading to v6 of slevomat/coding-standard, did that well, had all tests green and had 4 approvals, ready to be merged pending a few tests in our repos. That's not the correct moment to significantly expand the scope of this PR.

**Comment by lcobucci** [comment]
@alcaeus then we reach a deadlock... having it as `false` is a no-go from me and the current one is a no-go from you... let's wait for other people then :+1:

**Comment by alcaeus** [comment]
Having it as `false` is a sensible default for us, since almost all of our projects require PHP 7.1 or 7.2: even when running on PHP 7.4, the sniff may not require native type hints for us.

Projects that have PHP 7.4 as minimum version should change this setting to `true` and force native type hints.

**Comment by lcobucci** [comment]
@alcaeus just for consistency sake, that kind of "implicit behaviour" on sniffs happened in the past (like `object` type declaration and we're were good with it (even for projects that didn't require PHP 7.2 yet). I don't know that was fine back then and it's not anymore.

**Comment by alcaeus** [comment]
Then that slipped through the cracks, I'm also not happy with that decision: sniff behaviour should not depend on the PHP version that phpcs happens to be running on, but the minimum version of PHP required in composer.json.

**Comment by kukulich** [comment]
I think that false for native type hints for properties is current state before Slevomat CS 6.0

**Comment by lcobucci** [comment]
Okay, I'm back from my meeting and will try to clarify why having `enableNativeTypeHint=false` is a no-go from me.

The fact we don't have projects requiring PHP 7.4 is a different story (which is sad to me but I understand the reasons) but despite that this is a library used by other folks and it shouts to everybody (trend setting as @Ocramius keeps saying): **ALWAYS DECLARE THE FREAKING TYPES**.

The whole Makefile + `git patch` approach was the easiest way to ensure this message is still propagated (making sure that anyone can checkout this project and run `make` to test things properly), otherwise build fails on PHP 7.4 (without `enableNativeTypeHint=false`) for obvious reasons.

**Comment by alcaeus** [comment]
@lcobucci please document the process of maintaining the patch file for contributors and maintainers and I'm good with it. As is, a contributor would have to try and figure out how to fix the build, and I as maintainer would have to guess if I wanted to help them: this isn't how this stuff should be built. As previously stated, I can live with the setting being enabled depending on the PHP versions (especially since this is consistent with other settings) as long as the proper documentation is in place.

**Comment by lcobucci** [comment]
@alcaeus that's completely fair to ask, I'll handle that :+1:

**Comment by kukulich** [comment]
@lcobucci I think this PR is about upgrading Slevomat CS and false is compatible with previous behaviour. So it may be better to change the option in different PR

**Comment by ostrolucky** [comment]
All 7.4 builds will be red for PRs targetting our maintained branches, this is why it's bad. But like I said, let's talk about this on different place

**Comment by lcobucci** [comment]
> All 7.4 builds will be red for PRs targetting our maintained branches

They won't because it would be a new major release.

**Comment by alcaeus** [comment]
> All 7.4 builds will be red for PRs targetting our maintained branches (PHP >= 7.2 support), this is why it's bad. But like I said, let's talk about this on different place

To add on to what Luis said, most of our maintained branches run phpcs on the minimum required PHP version (as they should), not on PHP 7.4.

**Comment by lcobucci** [comment]
@alcaeus documentation added and make targets are now a bit more resilient.
